### PR TITLE
Implement the relative phase Toffoli gates

### PR DESCRIPTION
**Pull request by Cryoris**
### Summary

Implement the relative phase Toffoli gates (CCX and CCCX), which only existed as functions and not `Gate`s. These gates are used in the definition of other multi-controlled gates and are blocking their implementation.

### Details and comments

These gates follow https://arxiv.org/pdf/1508.03273.pdf. The relative-phase CCX is shown in Fig. 3, dashed box and the relative-phase CCCX is shown in Fig. 4, the whole circuit.

Things left to decide:
- [x] How should the gate class be called? The method used to be `rccx`, which conflicts with the naming scheme we introduce in #3633 where we use `R` to denote a rotation. 
**Edit:** Named `rccx/RCCXGate` since the method is already called `rccx`.
- [x] How should the relative phase Toffolis be drawn in the circuit drawer? Drawing them with `X` as base gate suggests that they are the same as the Toffolis -- which they are not. 
**Edit:** Resolved by not being a controlled gate, thus the name is drawn in a box (composite gate style).
- [x] Do these gates have a `base_gate`? The first idea might be to use the `XGate` as base gate, but this is not accurate, as the base gate implies that the controlled gate can be written as 
`controlled_gate = |0><0| \times id + |1><1| \times X`.
Also this runs into problems with the default mechanisms of Qiskit to create a controlled gate. 
If they don't have a `base_gate` this must be accordingly dealt with in the circuit drawer.
**Edit:** Resolved by not being a `ControlledGate` but `Gate` type for now.

### CODE REVIEW

**Comment by ajavadia** [comment]
We can just call it the Margolus gate? See https://github.com/Qiskit/qiskit-terra/issues/1741 (item 8) and https://arxiv.org/abs/quant-ph/0312225.

For now I think drawing it with a box and its name inside makes sense, unless we can come up with a modified notation for ccx that shows a relative phase difference. 

This is not a `ControlledGate`, and as such should not have `base_gate`. The matrix for it is (using big endian notation):
```
array([[ 1.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  1.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  1.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j,  1.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  1.+0.j,  0.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j, -1.+0.j,  0.+0.j,  0.+0.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.-1.j],
       [ 0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+0.j,  0.+1.j,  0.+0.j]])
```

**Comment by Cryoris** [comment]
Agreed to not having it as controlled gate, that solves the issue. 👍 
The matrix representation is also the one implemented in the class at the moment (with Qiskit's ordering).

However I'm not sure about Margolus gate, since (1) we tried to get consistent "action-based" gate names, and (2) how would we call the method then? 
In the paper you attached this gate is referred to as simplified Toffoli, what do you think about `SimplifiedCCXGate/sccx`? Or maybe keep the method name `rccx`, which is what its been up to now.

**Comment by ajavadia** [comment]
Ok I'm fine leaving rccx (that causes least disruption for now). I would at least put Margolus / Simplified Toffoli in the documentation because that seems a more common way of referring to this gate. 

There's some limit to "action-based" descriptions of gates, as gates have come to have common names in the literature and community. Like S gate and V gate, whose actions are basically rz90 and sqrt-x, but everyone just refers to them by the former names.

**Comment by Cryoris** [comment]
Ok, I left the method name as `rccx/rcccx` and named the classes `RCCXGate/RCCCXGate`. The docstrings describe them as simplified Toffoli or Margolus, or simplified 3-controlled Toffoli, respectively. If that naming is fine, the PR is ready to go from my side. 👍

File: qiskit/extensions/standard/rcccx.py [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    The simplified Toffoli gate implements the Toffoli gate up to relative phases.
+    Note, that the simplified Toffoli is not equivalent to the Toffoli. But can be used in places
+    where the Toffoli gate is uncomputed again.
+
+    This concrete implementation is from https://arxiv.org/abs/1508.03273, the complete circuit
+    of Fig. 4.
+    """
+
+    def __init__(self):
+        """Create a new PCCCX gate."""
```
Comment (ajavadia) [inline qiskit/extensions/standard/rcccx.py]:
```suggestion
        """Create a new RCCCX gate."""
```

File: qiskit/extensions/standard/rcccx.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            (CnotGate(), [q[2], q[3]], []),
+            (U1Gate(-pi / 4), [q[3]], []),
+            (U2Gate(0, pi), [q[3]], []),
+        ]
+        for inst in rule:
+            definition.append(inst)
+        self.definition = definition
+
+    def to_matrix(self):
+        """Return a numpy.array for the PCCCX gate."""
```
Comment (ajavadia) [inline qiskit/extensions/standard/rcccx.py]:
```suggestion
        """Return a numpy.array for the RCCCX gate."""
```

File: qiskit/extensions/standard/rcccx.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+                            [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0],
+                            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0],
+                            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, -1j, 0, 0, 0, 0],
+                            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0],
+                            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0],
+                            [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0],
+                            [0, 0, 0, 0, 0, 0, 0, -1, 0, 0, 0, 0, 0, 0, 0, 0]], dtype=complex)
+
+
+def rcccx(self, ctl1, ctl2, ctl3, tgt):
```
Comment (ajavadia) [inline qiskit/extensions/standard/rcccx.py]: These kwargs recently got changed to `control_qubit` and `target_qubit`, can you change them here for consistency?
See the table here: https://qiskit.org/documentation/release_notes.html#deprecation-notes
This was done in https://github.com/Qiskit/qiskit-terra/pull/3788

File: qiskit/extensions/standard/rccx.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+                            [0, 1, 0, 0, 0, 0, 0, 0],
+                            [0, 0, 1, 0, 0, 0, 0, 0],
+                            [0, 0, 0, 0, 0, 0, 0, -1j],
+                            [0, 0, 0, 0, 1, 0, 0, 0],
+                            [0, 0, 0, 0, 0, -1, 0, 0],
+                            [0, 0, 0, 0, 0, 0, 1, 0],
+                            [0, 0, 0, 1j, 0, 0, 0, 0]], dtype=complex)
+
+
+def rccx(self, ctl1, ctl2, tgt):
```
Comment (ajavadia) [inline qiskit/extensions/standard/rccx.py]: same for these kwargs

**Pull request merged**